### ISSUE TITLE
501 not implemented on GET.BUCKET

### ISSUE THREAD

**Issue by wzrdtales** [seaweedfs/seaweedfs#4853]
Happens for example when trying to use cockroachdb backup onto seaweedfs s3.

Audit log entry

```
2023-09-19T09:34:02+00:00	s3.access	{"request_id":"121940fcf63dff0c62972347c18bad2b","key":"/","bucket":"dt","status":501,"user_agent":"aws-sdk-go/1.40.37 (go1.19.10; linux; amd64)","remote_ip":"dt","operation":"REST.GET.BUCKET","host_id":"","host_header":"dt","time":1695116042,"error_code":"NotImplemented","signature_version":"SigV4","requester":"dt"}
```
We're looking for more information right now.

**Describe the bug**

Get Actions should work, but don't in this case.

**System Setup**

- weed 3.56

**Comment by wzrdtales** [comment]
i see the implementation is actually not according to the spec, probably to keep it simple and at least cover a little use case

**Comment by chrislusf** [comment]
the constant `ListingDelimDataSlash = "data/"` seems not used in this particular error.

need a simple way to reproduce this. Better if without using cockroachdb.

**Comment by wzrdtales** [comment]
no it is used, I already verified it is theerror.

```
 Name: 'test',
  Prefix: '',
  Delimiter: '/data/',
  MaxKeys: 1000,
  CommonPrefixes: [
    { Prefix: '2022/11/06-140519.79/data/' },
    { Prefix: '2022/11/30-105116.53/data/' },
    { Prefix: '2022/11/30-111234.48/data/' },
    { Prefix: '2022/11/30-145552.32/data/' },
    { Prefix: '2022/11/30-145604.30/data/' },
    { Prefix: '2023/01/02-084729.12/data/' },
    { Prefix: '2023/01/02-084750.71/data/' },
    { Prefix: '2023/01/03-010430.86/data/' },
    { Prefix: '2023/01/03-011812.49/data/' },
    { Prefix: '2023/01/03-011828.62/data/' },
    { Prefix: 'crdb-dtd/2023/07/16-160216.38/data/' },
    { Prefix: 'crdb-dtd/2023/07/21-124815.80/data/' },
    { Prefix: 'crdb-dtd/2023/07/26-162836.30/data/' },
    { Prefix: 'crdb-dtd/2023/07/26-163301.40/data/' },
    { Prefix: 'crdb-dtd/2023/09/19-102746.91/data/' },
    { Prefix: 'crdb/2023/06/20-084450.68/data/' },
    { Prefix: 'crdbgitdis/2023/06/20-100644.56/data/' },
    { Prefix: 'crdbgitdis/2023/06/20-102234.62/data/' },
    { Prefix: 'crdbgitdis/2023/06/20-104539.15/data/' }
  ]
```

this is what is expected to be returned.


And here my script I used to verify (node js you only need to npm init && npm install aws-sdk)

```js
const AWS = require('aws-sdk');
var ep = new AWS.Endpoint(process.env.AWS_ENDPOINT);
const s3 = new AWS.S3({
    region: process.env.AWS_REGION,
    accessKeyId: process.env.AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY,
    endpoint: ep,
  s3ForcePathStyle: true
});

async function getAllObjectsFromS3Bucket(bucket) {
    let isTruncated = true;
    let marker;
      let params = {Bucket: bucket, Delimiter: '/data/'};
        if (marker) params.Marker = marker;
        const response = await s3.listObjects(params).promise();
      console.log(response)
        
}

( async () => {

  console.log(await getAllObjectsFromS3Bucket('test'))
})()
```

**Comment by wzrdtales** [comment]
Here a more details, the delimiter is used here https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/backupdest/backup_destination.go#L493

And as stated before, I already tested with the script, that this indeed the issue.

**Comment by wzrdtales** [comment]
@chrislusf This is actually a major flaw in the implementation, which breaks the s3 protocol for a lot of tools out there. If you want to validate just one, cockroachdb backups would be one.

Do you think you may get some time to look at it?

**Comment by chrislusf** [comment]
This needs a tree traversal to find the matching entries, different from just directory listing.

**Comment by wzrdtales** [comment]
yeah, just asking since seaweed probably should follow the standard of s3 and in this case it deviates a lot and breaks therefor of course.

**Comment by wzrdtales** [comment]
https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html#API_ListObjectsV2_RequestSyntax

> CommonPrefixes contains all (if there are any) keys between Prefix and the next occurrence of the string specified by a delimiter.

To quote the official spec.

This is a quite heavyweight operation (as far as I know on AWS it actually incurs a lot of billed read actions). That is ok though for its purpose, it is still less heavyweight than manually recursing every path.

**Comment by wzrdtales** [comment]
Considering:

https://github.com/seaweedfs/seaweedfs/blob/53be97d5bea8b44db40077c62af5e224e21be3ee/weed/filer/postgres/postgres_sql_gen.go#L45

that might be actually possible by extending the functionality, to extend `prefix+%` with optionally `prefix+%+delimiter+%`.

While for the KV side of things, it requires more code. Outcome is the same, just easier to implement on the SQL side of things.

**Comment by wzrdtales** [comment]
Having not checked this in detail, is `cursor.maxKeys` decreasing per iteration, or per path that matched? It looks to me that it is the former, but would need to be the latter. At least that is what I remember when I took a look at the logic.

Logic wise per se, your implementation is up to the spec yes.

**Comment by kmlebedev** [comment]
You need to apply the matches locally and run c3 tests with the delimeter.
See https://github.com/seaweedfs/seaweedfs/pull/5237#issuecomment-1908771738

**Comment by wzrdtales** [comment]
cool, are you going to put a PR out for this?

**Comment by adasauce** [comment]
@kmlebedev thanks for pointing out the test suite, I had the implementation totally wrong according to ceph's suite.

I rebuilt the container using latest s3-tests because there many more delimiter based tests and the ceph s3-tests commit in the container is ~3years old.  The delimiter specific tests currently passing for me with my new implementation.

in general though, the seaweed s3 api is struggling with a few of the more basic prefix tests though, I'm not sure if it's in the scope of this to solve, but seems like a good place to look next.
